### PR TITLE
EVG-16035 insert a new intent for failed hosts

### PR DESCRIPTION
**Pull request by ybrill**
[EVG-16035](https://jira.mongodb.org/browse/EVG-16035)

### Description 
Currently, when a task spawns a host that AWS terminates immediately after giving it to us we fail the task on `host.list`.
This will instead insert a new host intent so we try again.

We will insert a new intent only if the original host is still "starting". If the host went to running and then terminated the task could have moved on and this would be a waste. I've never actually seen a "running" instance terminate unexpectedly, but just in case.
We also skip this if the task is not running or aborted because it would be a waste.

I'm undecided about limiting the number of times we loop on this. I didn't do it yet, but if someone feels strongly I'm game to add it. I think it's safe to assume it's a transient error since we know the instance was created once successfully (so it's not a user error). Even if the error does persist it won't loop infinitely because `host.list` will eventually reach its timeout.

### Testing 
I manually terminated a task's host in staging (very carefully!) and Evergreen created a second host for the task.

### CODE REVIEW

File: model/host/host.go [hunk 1]
```diff
@@ -1600,6 +1600,37 @@ func (h *Host) AddSSHKeyName(name string) error {
 	return nil
 }
 
+// HandleTerminatedHostSpawnedByTask should be called when a host spawned by a task
```
Comment (Kimchelly) [inline model/host/host.go:1603]: I would rephrase this doc to instead say something along the lines of "re-creates a new intent host if possible when this host.create host fails; otherwise if it cannot create a new host, it'll populate the reason that host.create failed".

File: model/host/host_intent.go [hunk 2]
```diff
@@ -78,6 +78,37 @@ func NewIntent(d distro.Distro, instanceName, provider string, options CreateOpt
 	return intentHost
 }
 
+func newIntentFromHost(h *Host) *Host {
```
Comment (Kimchelly) [inline model/host/host_intent.go:81]: `newIntentFromExisting`?

File: model/host/host_intent.go [hunk 3]
```diff
@@ -78,6 +78,37 @@ func NewIntent(d distro.Distro, instanceName, provider string, options CreateOpt
...
+	newID := h.Distro.GenerateName()
+	intentHost := &Host{
```
Comment (Kimchelly) [inline model/host/host_intent.go:83]: This struct initialization seems really similar to `NewIntent` in that both set the same fields in much the same way to achieve the same goal. Can we make both of them call the same common method to create the new host model to keep them in agreement on how to create an intent host?

File: model/host/host_test.go [hunk 4]
```diff
@@ -4829,6 +4829,112 @@ func TestSetNewSSHKeys(t *testing.T) {
...
+				SpawnOptions: SpawnOptions{
+					SpawnedByTask: false,
+				},
+			},
+			newIntentCreated: false,
+			hostDetailsSet:   false,
+		},
+	} {
+		t.Run(name, func(t *testing.T) {
+			require.NoError(t, db.ClearCollections(Collection, task.Collection))
```
Comment (Kimchelly) [inline model/host/host_test.go:4910]: I would defer clear the collections at the end of these tests to avoid leaving random test documents in people's DBs.

File: model/host/host_test.go [hunk 5]
```diff
@@ -4829,6 +4829,112 @@ func TestSetNewSSHKeys(t *testing.T) {
...
+			require.NoError(t, testCase.t.Insert())
+
+			assert.NoError(t, testCase.h.HandleTerminatedHostSpawnedByTask())
+
+			intent, err := FindOne(db.Query(nil))
+			require.NoError(t, err)
+			if testCase.newIntentCreated {
+				require.NotNil(t, intent)
+				assert.Contains(t, intent.Id, "evg")
```
Comment (Kimchelly) [inline model/host/host_test.go:4919]: `assert.True(t, host.IsIntentHostId(intent.Id))`

File: units/host_status.go [hunk 6]
```diff
@@ -185,8 +184,8 @@ func (j *cloudHostReadyJob) setCloudHostStatus(ctx context.Context, m cloud.Mana
 
 		catcher := grip.NewBasicCatcher()
 		if h.SpawnOptions.SpawnedByTask {
```
Comment (Kimchelly) [inline units/host_status.go:186]: It seems like checking if it's spawned by a task is redundant if it's already checked in the function.

File: units/host_monitoring_external_termination.go [hunk 7]
```diff
@@ -134,16 +133,18 @@ func handleExternallyTerminatedHost(ctx context.Context, id string, env evergree
 		if cloudStatus != cloud.StatusTerminated && (h.UserHost || h.StartedBy != evergreen.User) {
 			return false, errors.New("non-agent host is not already terminated and should not be terminated")
 		}
+
 		if h.SpawnOptions.SpawnedByTask {
```
Comment (Kimchelly) [inline units/host_monitoring_external_termination.go:137]: It seems like checking if it's spawned by a task is redundant if it's already checked in the function.

File: model/host/host_test.go [hunk 8]
```diff
@@ -4829,6 +4829,112 @@ func TestSetNewSSHKeys(t *testing.T) {
...
+					SpawnedByTask:       true,
+					TaskID:              "t0",
+					TaskExecutionNumber: 0,
+				},
+				Status: evergreen.HostStarting,
+			},
+			newIntentCreated: false,
+			hostDetailsSet:   true,
+		},
+		"HostAlreadyRuning": {
```
Comment (Kimchelly) [inline model/host/host_test.go:4876]: Running

File: model/host/host.go [hunk 9]
```diff
@@ -1600,6 +1600,37 @@ func (h *Host) AddSSHKeyName(name string) error {
...
+		intent := newIntentFromHost(h)
+		grip.Info(message.Fields{
+			"message":              "inserting a host intent to replace a terminated host for a task",
+			"original_host":        h.Id,
+			"original_host_status": h.Status,
+			"new_host":             intent.Id,
+			"task":                 t.Id,
+			"task_execution":       t.Execution,
+		})
+		return intent.Insert()
```
Comment (Kimchelly) [inline model/host/host.go:1628]: It seems like if this insertion fails, host.create will be considered a failure since it won't get another host. It seems like this should also call `AddHostCreateDetails` in the insertion error case.

Same with the other error cases above this - if they're not transient errors and this method does not retry (at the Amboy job level), it seems like they should populate the host create details.

File: model/host/host.go [hunk 10]
```diff
@@ -1600,6 +1600,37 @@ func (h *Host) AddSSHKeyName(name string) error {
...
+	}
+	if t == nil {
+		return errors.Errorf("host '%s' created by task '%s' execution '%d' that does not exist", h.Id, h.SpawnOptions.TaskID, h.SpawnOptions.TaskExecutionNumber)
+	}
+
+	if h.Status == evergreen.HostStarting && t.Status == evergreen.TaskStarted && !t.Aborted {
+		intent := newIntentFromHost(h)
+		grip.Info(message.Fields{
+			"message":              "inserting a host intent to replace a terminated host for a task",
+			"original_host":        h.Id,
```
Comment (Kimchelly) [inline model/host/host.go:1622]: nit: Most of the code base use `"host_id"` to reference fields containing host ID fields in grip logs (none of the other DB models include `_id`, but hosts do). Can we also make it consistent here, so it's `"original_host_id"` and `"new_host_id"`?

**Review by Kimchelly — CHANGES_REQUESTED** [review]

**Comment by ablack12** [comment]
> I'm undecided about limiting the number of times we loop on this. I didn't do it yet, but if someone feels strongly I'm game to add it. I think it's safe to assume it's a transient error since we know the instance was created once successfully (so it's not a user error). Even if the error does persist it won't loop infinitely because host.list will eventually reach its timeout.

I would like to press to limit the number of times we do this. Part of why we have host.list fail intentionally (with information) is so that we can present the user (and ourselves) with information. If we transiently fail even three times, I would guess something larger is wrong with AWS, and we could say something to the user to that effect.

File: model/host/host.go [hunk 11]
```diff
@@ -1600,6 +1600,37 @@ func (h *Host) AddSSHKeyName(name string) error {
 	return nil
 }
 
+// HandleTerminatedHostSpawnedByTask should be called when a host spawned by a task
+// is found terminated.
+func (h *Host) HandleTerminatedHostSpawnedByTask() error {
```
[the remark this answers is not in the document]
Comment (ybrill) [reply model/host/host.go]: moved into the host-monitoring-external-state-check job

File: model/host/host.go (hunk 1 above)
Comment (ybrill) [reply model/host/host.go]: done (though now it's private)

File: model/host/host_intent.go (hunk 2 above)
Comment (ybrill) [reply model/host/host_intent.go]: this went away

File: model/host/host_intent.go (hunk 3 above)
Comment (ybrill) [reply model/host/host_intent.go]: I think this is good now.

File: model/host/host_test.go (hunk 4 above)
Comment (ybrill) [reply model/host/host_test.go]: done

File: model/host/host_test.go (hunk 5 above)
Comment (ybrill) [reply model/host/host_test.go]: done

File: units/host_status.go (hunk 6 above)
Comment (ybrill) [reply units/host_status.go]: removed the extra check

File: model/host/host_test.go (hunk 8 above)
Comment (ybrill) [reply model/host/host_test.go]: fixed

File: model/host/host.go (hunk 9 above)
Comment (ybrill) [reply model/host/host.go]: I think I've done this now. 
Though if the db call to get the task fails perhaps that's not transient and we could retry 🤷

File: model/host/host.go (hunk 10 above)
Comment (ybrill) [reply model/host/host.go]: done

**Comment by ybrill** [comment]
@ablack12
> I would like to press to limit the number of times we do this

This is done. I made it retry twice (so we get three tries).

File: model/host/host.go [hunk 12]
```diff
@@ -309,6 +309,10 @@ type SpawnOptions struct {
 	// Retries is the number of times Evergreen should try to spawn this host.
 	Retries int `bson:"retries,omitempty" json:"retries,omitempty"`
 
+	// Respawns is the number of spawns remaining if the host is externally terminated after
```
Comment (Kimchelly) [inline model/host/host.go:312]: number of spawn attempts remaining

File: units/host_monitoring_external_termination.go [hunk 13]
```diff
@@ -168,3 +167,61 @@ func handleExternallyTerminatedHost(ctx context.Context, id string, env evergree
 		return false, errors.Errorf("unexpected host status '%s'", cloudStatus)
 	}
 }
+
+// handleTerminatedHostSpawnedByTask re-creates a new intent host if possible when this host.create host fails;
```
Comment (Kimchelly) [inline units/host_monitoring_external_termination.go:171]: I would put a period at the end of this sentence, or start L172 without capitalization.

**Review by Kimchelly — APPROVED** [review]
LGTM mod 2 minor comments.

Should the wiki be updated to mention that host.create will retry and propagate failed attempt reasons to the user?

File: units/host_monitoring_external_termination.go (hunk 13 above)
Comment (ybrill) [reply units/host_monitoring_external_termination.go]: done

File: model/host/host.go (hunk 12 above)
Comment (ybrill) [reply model/host/host.go]: done

**Comment by Kimchelly** [comment]
It seems like host.create/host.list failure reasons are an end user feature since we propagate internal failure reasons to host.list, which they have to be able to meaningfully interpret without resorting to #evergreen-users. I don't think that they need to know how many retries they're given, just let them know that they should expect host.list to include details for each spawn attempt failure.

**Comment by ybrill** [comment]
I'm confused by this: I thought host.list will only tell them "host was externally terminated" once when it's run out of retries because even if we retry twice we will still only add host details once.

**Comment by Kimchelly** [comment]
I didn't really look at host.list that carefully, but it seems like each host.create attempt failure tries to call `AddHostCreateDetails` and each call to `AddHostCreateDetails` pushes another error string to the details field. If host.list only exposes the last error in the host details, then it's probably fine, but I would otherwise expect them to receive error reasons for all host.create failed attempts.

**Comment by ablack12** [comment]
Pushing a host create details to a task implies that that host has completely failed; in host list, we count everyHostCreateDetail object as one of the hosts we're waiting for. I haven't read this code closely but we shouldn't add a HostCreateDetail for every error, only if it all fails at the end.

**Comment by ybrill** [comment]
That's definitely what I was aiming for: when this case occurs we either insert a new intent or push a HostCreateDetail.
Never both.

**Comment by ybrill** [comment]
@ablack12 and/or @Kimchelly do you intent to look at this more?
(if not I'll just merge)

**Comment by ablack12** [comment]
I'm good with it!

**Comment by Kimchelly** [comment]
SGTM
